File: web/src/core/hole.ts
// Hole solver — ported from Organon.Syl.Hole

import { moodSpec, validMoods } from "./tradition";
import { Figure, Mood, PropType, Term, Tradition, termEq } from "./types";

export type TermH =
  | { readonly tag: "ConcreteT"; readonly term: Term }
  | { readonly tag: "HoleT" };

export type PropTypeH =
  | { readonly tag: "ConcretePT"; readonly propType: PropType }
  | { readonly tag: "HolePT" };

export type PropositionH =
  | {
      readonly tag: "PropH";
      readonly propTypeH: PropTypeH;
      readonly subject: TermH;
      readonly predicate: TermH;
    }
  | { readonly tag: "WholePropH" };

export type SyllogismH = {
  readonly majorH: PropositionH;
  readonly minorH: PropositionH;
  readonly conclusionH: PropositionH;
};

export type SolutionProp = {
  readonly solPropType: PropType;
  readonly solSubject: Term | null;
  readonly solPredicate: Term | null;
};

export type Solution = {
  readonly solutionMood: Mood;
  readonly solMajor: SolutionProp;
  readonly solMinor: SolutionProp;
  readonly solConclusion: SolutionProp;
};

enum Role {
  S = "S",
  P = "P",
  M = "M",
}

type RolePair = [Role, Role];
type FigureRoles = [RolePair, RolePair, RolePair];

function figureRoles(fig: Figure): FigureRoles {
  switch (fig) {
    case Figure.FigI:
      return [
        [Role.M, Role.P],
        [Role.S, Role.M],
        [Role.S, Role.P],
      ];
    case Figure.FigII:
      return [
        [Role.P, Role.M],
        [Role.S, Role.M],
        [Role.S, Role.P],
      ];
    case Figure.FigIII:
      return [
        [Role.M, Role.P],
        [Role.M, Role.S],
        [Role.S, Role.P],
      ];
    case Figure.FigIV:
      return [
        [Role.P, Role.M],
        [Role.M, Role.S],
        [Role.S, Role.P],
      ];
  }
}

type Bindings = Map<Role, Term>;

function matchTerm(th: TermH, role: Role, bindings: Bindings): Bindings | null {
  if (th.tag === "HoleT") return bindings;
  const existing = bindings.get(role);
  if (existing === undefined) {
    const next = new Map(bindings);
    next.set(role, th.term);
    return next;
  }
  return termEq(existing, th.term) ? bindings : null;
}

function matchProp(
  propH: PropositionH,
  expectedPT: PropType,
  subjR: Role,
  predR: Role,
  bindings: Bindings,
): Bindings | null {
  if (propH.tag === "WholePropH") return bindings;
  if (
    propH.propTypeH.tag === "ConcretePT" &&
    propH.propTypeH.propType !== expectedPT
  )
    return null;
  const b1 = matchTerm(propH.subject, subjR, bindings);
  if (b1 === null) return null;
  return matchTerm(propH.predicate, predR, b1);
}

export function solve(tradition: Tradition, sylH: SyllogismH): Solution[] {
  const results: Solution[] = [];
  for (const mood of validMoods(tradition)) {
    const spec = moodSpec(mood);
    const fig = spec.moodFigure;
    const [majRoles, minRoles, conRoles] = figureRoles(fig);

    const items: [PropositionH, PropType, Role, Role][] = [
      [sylH.majorH, spec.majorPropType, majRoles[0], majRoles[1]],
      [sylH.minorH, spec.minorPropType, minRoles[0], minRoles[1]],
      [sylH.conclusionH, spec.conclusionPropType, conRoles[0], conRoles[1]],
    ];

    let bindings: Bindings | null = new Map();
    for (const [propH, expectedPT, subjR, predR] of items) {
      bindings = matchProp(propH, expectedPT, subjR, predR, bindings);
      if (bindings === null) break;
    }

    if (bindings !== null) {
      const resolve = (r: Role): Term | null => bindings!.get(r) ?? null;
      results.push({
        solutionMood: mood,
        solMajor: {
          solPropType: spec.majorPropType,
          solSubject: resolve(majRoles[0]),
          solPredicate: resolve(majRoles[1]),
        },
        solMinor: {
          solPropType: spec.minorPropType,
          solSubject: resolve(minRoles[0]),
          solPredicate: resolve(minRoles[1]),
        },
        solConclusion: {
          solPropType: spec.conclusionPropType,
          solSubject: resolve(conRoles[0]),
          solPredicate: resolve(conRoles[1]),
        },
      });
    }
  }
  return results;
}

File: src/Organon/Syl/Hole.hs
module Organon.Syl.Hole
  ( TermH (..),
    PropTypeH (..),
    PropositionH (..),
    SyllogismH (..),
    SolutionProp (..),
    Solution (..),
    solve,
  )
where

import qualified Data.Map.Strict as Map
import Organon.Syl.Tradition
import Organon.Syl.Types

-- | A term that may be a hole.
data TermH = ConcreteT Term | HoleT
  deriving stock (Eq, Show)

-- | A proposition type that may be a hole.
data PropTypeH = ConcretePT PropType | HolePT
  deriving stock (Eq, Show)

-- | A proposition that may contain holes at any level,
-- or be entirely unknown.
data PropositionH
  = PropH PropTypeH TermH TermH
  | WholePropH
  deriving stock (Eq, Show)

-- | A syllogism with possible holes.
data SyllogismH = SylH
  { majorH :: PropositionH,
    minorH :: PropositionH,
    conclusionH :: PropositionH
  }
  deriving stock (Eq, Show)

-- | A proposition in a solution, where terms may be unresolved.
-- 'Nothing' means the role was unconstrained (a meta-variable).
data SolutionProp = SolutionProp
  { solPropType :: PropType,
    solSubject :: Maybe Term,
    solPredicate :: Maybe Term
  }
  deriving stock (Eq, Show)

-- | A solution: a mood and the (possibly partially-filled) syllogism.
data Solution = Solution
  { solutionMood :: Mood,
    solMajor :: SolutionProp,
    solMinor :: SolutionProp,
    solConclusion :: SolutionProp
  }
  deriving stock (Eq, Show)

-- Term roles in a syllogism.
data Role = RoleS | RoleP | RoleM
  deriving stock (Eq, Ord, Show)

-- | The role assignments for each term position given a figure.
figureRoles :: Figure -> ((Role, Role), (Role, Role), (Role, Role))
figureRoles FigI = ((RoleM, RoleP), (RoleS, RoleM), (RoleS, RoleP))
figureRoles FigII = ((RoleP, RoleM), (RoleS, RoleM), (RoleS, RoleP))
figureRoles FigIII = ((RoleM, RoleP), (RoleM, RoleS), (RoleS, RoleP))
figureRoles FigIV = ((RoleP, RoleM), (RoleM, RoleS), (RoleS, RoleP))

-- | Find all valid syllogisms that match the given pattern with holes.
solve :: Tradition -> SyllogismH -> [Solution]
solve tradition (SylH majH minH conH) =
  concatMap tryMood (validMoods tradition)
  where
    tryMood mood =
      let spec = moodSpec mood
          fig = moodFigure spec
          ((majSR, majPR), (minSR, minPR), (conSR, conPR)) = figureRoles fig
       in case matchAll
            [ (majH, majorPropType spec, majSR, majPR),
              (minH, minorPropType spec, minSR, minPR),
              (conH, conclusionPropType spec, conSR, conPR)
            ]
            Map.empty of
            Nothing -> []
            Just bindings ->
              let resolve r = Map.lookup r bindings
               in [ Solution
                      mood
                      (SolutionProp (majorPropType spec) (resolve majSR) (resolve majPR))
                      (SolutionProp (minorPropType spec) (resolve minSR) (resolve minPR))
                      (SolutionProp (conclusionPropType spec) (resolve conSR) (resolve conPR))
                  ]

    matchAll [] bindings = Just bindings
    matchAll ((propH, expectedPT, subjR, predR) : rest) bindings =
      case matchProp propH expectedPT subjR predR bindings of
        Nothing -> Nothing
        Just bindings' -> matchAll rest bindings'

    matchProp WholePropH _ _ _ bindings = Just bindings
    matchProp (PropH ptH subjH predH) expectedPT subjR predR bindings =
      case ptH of
        ConcretePT pt | pt /= expectedPT -> Nothing
        _ -> matchTerm subjH subjR bindings >>= matchTerm predH predR

    matchTerm HoleT _ bindings = Just bindings
    matchTerm (ConcreteT t) role bindings =
      case Map.lookup role bindings of
        Nothing -> Just (Map.insert role t bindings)
        Just t'
          | t == t' -> Just bindings
          | otherwise -> Nothing
